File: src/main/java/akka/actor/ReduceActor.java
package akka.actor;

import akka.message.MapDataResult;
import akka.message.ReduceDataResult;
import akka.message.WordCount;

import java.util.HashMap;
import java.util.List;

/**
 * @user sun
 * @date 2016/7/4
 */
public class ReduceActor extends UntypedActor {

    private ActorRef aggregateActor;

    public ReduceActor(ActorRef aggregateActor) {
        this.aggregateActor = aggregateActor;
    }

    @Override
    public void onReceive(Object message) throws Exception {
        if (message instanceof MapDataResult) {
            MapDataResult dataResult = (MapDataResult) message;
            ReduceDataResult result = reduce(dataResult.getDataList());
            aggregateActor.tell(result, getSelf());
        } else
            unhandled(message);
    }

    private ReduceDataResult reduce(List<WordCount> wordCounts) {
        HashMap<String, Integer> reduceMap = new HashMap<>();
        for (WordCount wordCount : wordCounts) {
            if (reduceMap.containsKey(wordCount.getWord())) {
                int value = reduceMap.get(wordCount.getWord());
                value++;
                reduceMap.put(wordCount.getWord(), value);
            } else {
                reduceMap.put(wordCount.getWord(), 1);
            }
        }

        return new ReduceDataResult(reduceMap);
    }
}

File: src/main/scala/scala_akka/actor/ReduceActor.scala
package scala_akka.actor

import java.util

import akka.actor.{Actor, ActorRef}

import scala_akka.message.{MapDataResult, ReduceDataResult, Word}

/**
  * @user sun 
  * @date 2016/7/10
  */
class ReduceActor(aggregateActor: ActorRef) extends Actor {
  override def receive: Receive = {
    case message: MapDataResult =>
      aggregateActor ! reduce(message.dataList)
    case _ =>
  }

  def reduce(dataList: util.ArrayList[Word]): ReduceDataResult = {
    val reduceMap = new util.HashMap[String, Integer]
    import scala.collection.JavaConversions._
    for (ws: Word <- dataList) {
      if (reduceMap.containsKey(ws.word)) {
        var value: Int = reduceMap.get(ws.word)
        value += 1
        reduceMap.put(ws.word, value)
      }
      else {
        reduceMap.put(ws.word, 1)
      }
    }
    new ReduceDataResult(reduceMap)
  }
}
